core/unix/signal_linux_x86.c

```diff
@@ -237,20 +260,44 @@ save_xmm(dcontext_t *dcontext, sigframe_rt_t *frame)
         if (YMM_ENABLED()) {
             /* i#637: ymm top halves are inside kernel_xstate_t */
             memcpy(&xstate->ymmh.ymmh_space[i * 4],
-                   ((void *)&get_mcontext(dcontext)->simd[i]) + XMM_REG_SIZE,
+                   ((void *)&get_mcontext(dcontext)->simd[i]) + YMMH_REG_SIZE,
                    YMMH_REG_SIZE);
         }
 #else
         memcpy(&sc->fpstate->_xmm[i], &get_mcontext(dcontext)->simd[i], XMM_REG_SIZE);
         if (YMM_ENABLED()) {
             /* i#637: ymm top halves are inside kernel_xstate_t */
             memcpy(&xstate->ymmh.ymmh_space[i * 4],
-                   ((void *)&get_mcontext(dcontext)->simd[i]) + XMM_REG_SIZE,
+                   ((void *)&get_mcontext(dcontext)->simd[i]) + YMMH_REG_SIZE,
                    YMMH_REG_SIZE);
         }
+#endif
+#ifdef X64
+        if (ZMM_ENABLED()) {
+            memcpy((byte *)xstate + proc_xstate_area_zmm_hi256_offs() + i * ZMMH_REG_SIZE,
+                   ((void *)&get_mcontext(dcontext)->simd[i]) + ZMMH_REG_SIZE,
+                   ZMMH_REG_SIZE);
+            memcpy((byte *)xstate + proc_xstate_area_hi16_zmm_offs() + i * ZMM_REG_SIZE,
+                   ((void *)&get_mcontext(dcontext)
+                        ->simd[i + proc_num_simd_sse_avx_saved()]),
```

Comment:
Similar comment to above: maybe this should be in a separate loop.  At least an assert on the count.

Reply:
Assert added.